src/authorizer.ts

```diff
@@ -24,68 +27,61 @@ export interface RuleContext<User, Data = unknown> extends LoaderArgs {
  * Inside a Rule function you can do any validation to verify a user to continue
  * and return a promise resolving to a boolean value.
  */
-export interface RuleFunction<User, Data = unknown> {
+export interface RuleFunction<User, Data = null> {
   (context: RuleContext<User, Data>): Promise<boolean>;
 }
 
+type AuthorizeOptionsError = {
+  failureRedirect?: never;
+  raise: "error";
+};
+type AuthorizeOptionsResponse = {
+  failureRedirect?: never;
+  raise: "response";
+};
+type AuthorizeOptionsRedirect = {
+  failureRedirect: string;
+  raise: "redirect";
+};
+type AuthorizeOptionsEmpty = {
+  failureRedirect?: never;
+  raise?: never;
+};
+type AuthorizeOptions<U, D> = (
+  | AuthorizeOptionsError
+  | AuthorizeOptionsRedirect
+  | AuthorizeOptionsResponse
+  | AuthorizeOptionsEmpty
+) & { rules?: RuleFunction<U, D>[] };
+
 export class Authorizer<User = unknown, Data = unknown> {
   constructor(
     private authenticator: Authenticator<User>,
     private rules: RuleFunction<User, Data>[] = []
   ) {}
 
-  async authorize(
-    args: AuthorizeArgs<User, Data>,
-    options?: {
-      failureRedirect?: never;
-      raise?: "error";
-      rules?: RuleFunction<User, Data>[];
-    }
-  ): Promise<User>;
-  async authorize(
-    args: AuthorizeArgs<User, Data>,
-    options?: {
-      failureRedirect?: never;
-      raise?: "response";
-      rules?: RuleFunction<User, Data>[];
-    }
-  ): Promise<User>;
-  async authorize(
-    args: AuthorizeArgs<User, Data>,
-    options: {
-      failureRedirect: string;
-      raise: "redirect";
-      rules?: RuleFunction<User, Data>[];
+  async authorize<D extends Data>(
+    args: AuthorizeArgs<D>,
+    { failureRedirect, raise, rules = [] }: AuthorizeOptions<User, D> = {
+      raise: "response",
+      rules: [],
     }
-  ): Promise<User>;
-  async authorize(
-    args: AuthorizeArgs<User, Data>,
-    {
-      failureRedirect,
-      raise = "response",
-      rules = [],
-    }: {
-      failureRedirect?: string;
-      raise?: "error" | "response" | "redirect";
-      rules?: RuleFunction<User, Data>[];
-    } = {}
   ): Promise<User> {
+    if (!raise) raise = "response";
```

Comment:
I'm thinking the default raise should be error, I think that's more expected and if the developer setup Sentry it should know if this throw and the dev is not catching it.

Also, Authorizer may receive the default raise and use it as default here